File: sdks/php/src/WebhookVerifier.php
<?php

declare(strict_types=1);

namespace Jobcelis;

/**
 * Verify webhook signatures using HMAC-SHA256 (Base64, no padding).
 */
class WebhookVerifier
{
    /**
     * Verify that a webhook request body matches the provided HMAC-SHA256 signature.
     *
     * @param string $secret  The webhook signing secret.
     * @param string $body    The raw request body.
     * @param string $signature The signature from the X-Signature header (format: "sha256=<base64>").
     * @return bool True if the signature is valid.
     */
    public static function verify(string $secret, string $body, string $signature): bool
    {
        $prefix = 'sha256=';
        if (!str_starts_with($signature, $prefix)) {
            return false;
        }

        $receivedSig = substr($signature, strlen($prefix));
        $hash = hash_hmac('sha256', $body, $secret, true);
        $expected = rtrim(base64_encode($hash), '=');

        return hash_equals($expected, $receivedSig);
    }
}

File: sdks/dart/lib/src/webhook_verifier.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// Verifies webhook signatures using HMAC-SHA256 (Base64, no padding) with constant-time comparison.
///
/// Use [WebhookVerifier.verify] to validate incoming webhook requests.
class WebhookVerifier {
  /// Private constructor -- use the static [verify] method.
  WebhookVerifier._();

  static const String _prefix = 'sha256=';

  /// Verify a webhook signature.
  ///
  /// The [signature] should be the full X-Signature header value (format: "sha256=<base64>").
  /// Returns `true` if the signature is valid.
  static bool verify(String secret, String body, String signature) {
    if (!signature.startsWith(_prefix)) return false;

    final receivedSig = signature.substring(_prefix.length);
    final hmac = Hmac(sha256, utf8.encode(secret));
    final digest = hmac.convert(utf8.encode(body));
    final expected = base64Encode(digest.bytes).replaceAll('=', '');

    if (expected.length != receivedSig.length) return false;

    var result = 0;
    for (var i = 0; i < expected.length; i++) {
      result |= expected.codeUnitAt(i) ^ receivedSig.codeUnitAt(i);
    }
    return result == 0;
  }
}
